src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledArrayBuilder`1.cs

```diff
@@ -74,60 +193,125 @@ public void AddRange(ImmutableArray<T> items)
             return;
         }
 
-        _builder ??= GetBuilder();
-        _builder.AddRange(items);
+        if (_builder is { } builder)
+        {
+            builder.AddRange(items);
+        }
+        else if (_inlineCount + items.Length <= InlineCapacity)
+        {
+            foreach (var item in items)
+            {
+                SetInlineElement(_inlineCount, item);
+                _inlineCount++;
+            }
+        }
+        else
+        {
+            MoveInlineItemsToBuilder();
+            _builder.AddRange(items);
+        }
     }
 
-    public void AddRange(IReadOnlyList<T> items)
+    public void AddRange(IEnumerable<T> items)
     {
-        if (items.Count == 0)
+        if (!items.TryGetCount(out var count))
         {
+            // We can't retrieve a count, so we have to enumerate the elements.
+            AddRangeSlow(this, items);
             return;
         }
 
-        _builder ??= GetBuilder();
-        _builder.AddRange(items);
+        if (count == 0)
+        {
+            return;
+        }
+
+        if (_builder is { } builder)
```

Comment:
Feels like we could move this to the first thing here, as we do it regardless of whether `items` is countable or not.

Reply:
Yeah, I think this can be refactored a bit. Thanks!

Reply:
I tweaked it a bit. I don't love it, but it's better than it was before